File: Algorithms/Graphs/BFS.py

def _in_bounds(grid, point):

    r, c = point
    return min(r, c) >= 0 and r < len(grid) and c < len(grid[0])

def _get_neighbors(grid, point):

    r, c = point
    for rr, cc in [[r + 1, c], [r - 1, c], [r, c + 1], [r, c - 1]]:
        if _in_bounds(grid, (rr, cc)) and grid[rr][cc] == '.':
            yield rr, cc

def solve(grid, start, end):

    visited = []
    parents = {}

    q = [start]

    while len(q) > 0:

        cur = q.pop(0)
        visited.append(cur)

        if cur == end:

            path = []
            node = cur

            while node != None:
                path.insert(0, node)
                node = parents.get(node)

            return path

        else:

            for n in _get_neighbors(grid, cur):
                if n not in visited:
                    q.append(n)
                    parents.update({n : cur})

    return None


def solve2(grid, end, visited=None, parents=None, q=None, start=None):

    if visited == None or parents == None or q == None:

        q = [start]

        return solve2(grid, end, [], {}, q)

    elif len(q) == 0:
        return None

    cur = q.pop(0)

    visited.append(cur)

    if cur == end:

        path = []
        node = cur

        while node != None:
            path.insert(0, node)
            node = parents.get(node)

        return path

    else:

        for n in _get_neighbors(grid, cur):

            if n not in visited:

                parents.update({n : cur})

                q.append(n)

                new_path = solve2(grid, end, visited, parents, q)

                if new_path:

                    return new_path

    return None

def main():

    grid = [
        ".####",
        ".#...",
        ".#.#.",
        "...#.",
        ".##.."
    ]

    for p in solve(grid, (0,0), (4,4)):
        print(p)
    print("------")
    for p in solve2(grid, (4,4), start=(0,0)):
        print(p)

if __name__ == '__main__':
    main()

File: Algorithms/Graphs/BFS.js
"use strict";

Set.prototype.hasPoint = function(point) {

    for (let item of this.values()) {

        if (item.x == point.x && item.y == point.y) {

            return true;

        }

    }

    return false;

};

let inBounds = (grid, p) => Math.min(p.x, p.y) >= 0 && p.x < grid.length && p.y < grid[0].length;

function* getNeighbors(grid, point) {

    let {x, y} = point;

    for (let n of [{x: x + 1, y: y}, {x: x - 1, y: y}, {x: x, y: y + 1}, {x: x, y: y - 1}]) {

        if (inBounds(grid, n) && grid[n.x][n.y] == '.') {

            yield n;

        }

    }

}

let solve = (grid, start, end) => {

    let visited = new Set(),
        parents = new Map(),
        q = [start];

    while (q.length > 0) {

        let cur = q.shift();
        visited.add(cur);

        if (cur.x == end.x && cur.y == end.y) {

            let path = [],
                node = cur;

            while (node != null) {

                path.unshift(node);
                node = parents.get(node);

            }

            return path;

        } else {

            for (let n of getNeighbors(grid, cur)) {

                if (!visited.hasPoint(n)) {

                    q.push(n);
                    parents.set(n, cur);

                }

            }

        }

    }

    return null;

}

let solve2 = (grid, end, start = null, visited = new Set(), parents = new Map(), q = null) => {

    if (start != null && q == null) {

        q = [start];

    } else if (q.length == 0) {

        return null;

    }

    let cur = q.shift();
    visited.add(cur);

    if (cur.x == end.x && cur.y == end.y) {

        let path = [],
            node = cur;

        while (node != null) {

            path.unshift(node);
            node = parents.get(node);

        }

        return path;

    } else {

        for (let n of getNeighbors(grid, cur)) {

            if (!visited.hasPoint(n)) {

                q.push(n);
                parents.set(n, cur);

                let newPath = solve2(grid, end, null, visited, parents, q);

                if (newPath) {

                    return newPath;

                }

            }

        }

    }

}

let grid = [
    ".####",
    ".#...",
    ".#.#.",
    "...#.",
    ".##.."
];

for (let point of solve(grid, {x: 0, y: 0}, {x: 4, y: 4})) {

    console.log(point);

}

console.log("------");

for (let point of solve2(grid, {x: 4, y: 4}, {x: 0, y: 0})) {

    console.log(point);

}
